scaleway/resource_apple_silicon_server.go

```diff
@@ -0,0 +1,178 @@
+			"deletable_at": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The minimal date and time on which you can delete this server due to Apple licence",
+			},
+
+			// Common
+			"zone":            zoneSchema(),
+			"organization_id": organizationIDSchema(),
+			"project_id":      projectIDSchema(),
+		},
+	}
+}
+
+func resourceScalewayAppleSiliconServerCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	asAPI, zone, err := asAPIWithZone(d, m)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	createReq := &applesilicon.CreateServerRequest{
+		Name:      expandOrGenerateString(d.Get("name"), "m1"),
+		Type:      d.Get("type").(string),
+		ProjectID: d.Get("project_id").(string),
+	}
+
+	res, err := asAPI.CreateServer(createReq, scw.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(newZonedIDString(zone, res.ID))
+
+	_, err = asAPI.WaitForServer(&applesilicon.WaitForServerRequest{
+		ServerID:      res.ID,
+		Timeout:       scw.TimeDurationPtr(defaultAppleSiliconServerTimeout),
+		RetryInterval: nil,
+	}, scw.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceScalewayRdbInstanceRead(ctx, d, m)
+}
+
+func resourceScalewayAppleSiliconServerRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	asAPI, zone, ID, err := asAPIWithZoneAndID(m, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	res, err := asAPI.GetServer(&applesilicon.GetServerRequest{
+		Zone:     zone,
+		ServerID: ID,
+	}, scw.WithContext(ctx))
+	if err != nil {
+		if is404Error(err) {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	_ = d.Set("name", res.Name)
+	_ = d.Set("type", res.Type)
+	_ = d.Set("created_at", res.CreatedAt.Format(time.RFC3339))
+	_ = d.Set("updated_at", res.UpdatedAt.Format(time.RFC3339))
+	_ = d.Set("deletable_at", res.DeletableAt.Format(time.RFC3339))
+	_ = d.Set("ip", res.IP.String())
+	_ = d.Set("vnc_url", res.VncURL)
+
+	_ = d.Set("zone", zone.String())
+	_ = d.Set("organization_id", res.OrganizationID)
+	_ = d.Set("project_id", res.ProjectID)
+
+	return nil
+}
+
+func resourceScalewayAppleSiliconServerUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	asAPI, zone, ID, err := asAPIWithZoneAndID(m, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	req := &applesilicon.UpdateServerRequest{
+		Zone:     zone,
+		ServerID: ID,
+	}
+
+	if d.HasChange("name") {
+		req.Name = expandStringPtr(d.Get("name"))
+	}
+
+	_, err = asAPI.UpdateServer(req, scw.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceScalewayAppleSiliconServerRead(ctx, d, m)
+}
+
+func resourceScalewayAppleSiliconServerDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	asAPI, zone, ID, err := asAPIWithZoneAndID(m, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
```

Comment:
nit: we could look at the `deletable_at` value from the state and throw an error if it's not deletable.. but it might introduce bugs if the client doesn't have a good clock 🤔

Reply:
Yes I hesitate about this feature because it relies on an accurate clock that could be wrong. I think the safest approach is to code a complete WaitForPossibleDeletion function that would behave like a Wait and try periodically.

Reply:
https://github.com/scaleway/scaleway-sdk-go/issues/722